BAU - Rework log fetching for better automation compatibility

alexandergeegds opened the pull request:
Do not merge until:

- [x] #2 is merged

Includes some changes and (hopefully) improvements to the `fetch_missing_matomo_requests.py` script with the intention of making it easier to use in automated workflows. This includes adding some more environment variables to control various aspects of the script, giving it logging, and adding bash scripts in the style of the other components in this repo for standardising how the Docker image can be built and run.

alexandergeegds committed:
Be consistent in use of quotes

Double quotes are now only being used for multi-line strings.

alexandergeegds committed:
Extract some constants; simplify the date input

The date format is used in a few places. Note that the filename
format has here been changed slightly to make the formats consistent.
The filename suffix is only used once, but never changes. Both are
now extracted as constants and elevated to the top of the script for
visibility. The name of a variable tracking the output filename has
also been lowercased so that it is not assumed to be a constant.

Additionally, the input variable called START_DATE had previously
required a timestamp. For simplicity, this is now reduced to a date
as the name suggests, and the timestamp is assumed to be the start of
the day.

alexandergeegds committed:
Switch to the logging library

Currently, messages are being printed. This is fine if all output is
always wanted. However, during automated runs, it is often helpful to
be verbose to assist with debugging any potential problems. So being
able to print at different levels can be beneficial. Instead of
printing based on whether or not a debug flag is provided, the
logging library allows all logs to be created the same way, but only
printed if an appropriate flag is specified. Thus developers only
need to see verbose output when they want, but pipelines can specify
an appropriate parameter for suitably verbose output every run.

Currently, the log level can be overridden by environment variables,
in a similar fashion to the existing script parameters.

alexandergeegds committed:
Add more logging

This should allow an automated platform to output more verbose
information about the steps it is taking, such as the periods it
requests, how many requests are written for each period, and notably
whether the script is hanging while waiting for a response to
complete. This last is especially important, because if nothing is
printed, it can be hard to know what has caused the script to stall.

alexandergeegds committed:
Include stack trace in log output

Switching to the exception logger, the logging library automatically
includes information from the exception that was raised at no further
effort. Which increases visibility on what the problem is, especially
during remedying misconfigurations.

alexandergeegds committed:
Simplify loop logic

Instead of looping for a day, for each of the days specified, it
should be equivalent to simply determine the entire length of the
period, and continue looping until all requests in that period have
been downloaded.

Simplifying the loop in this way allows us to eliminate both an
unnecessary level of nesting, and an additional final iteration that
required checking outside the loop and thus duplicated the code
sending the requests.

The size of the period to loop over has been extracted into a
constant to increase its visibility. This is a temporary change until
an environment variable is established for it.

alexandergeegds committed:
Make period width overridable

This should mean adjustments to the period width can happen either in
code (if the default value should be changed) or through environment
variables (if a single run of the script is safe to use non-default
values for).

Since it is now possible to provide large values, the warning about
receiving too many requests is now an error and aborts the script.
This is also better behaviour for automation, since the process
running the script should also fail, and the error should propagate.

alexandergeegds committed:
Add script to retrieve logs from a specified day

The included script should make automating this process simpler. It
passes on the appropriate environment variables to the python script.
Note that some optional variables have purposely not been named, but
Docker should pass any values set even without the explicit
assignment. Explicit assignments have been retained for some to use
the built-in error-if-not-set bash functionality. If bash cannot be
used for automation, this may need further refinement.

Note also that both local helper scripts now also ensure that the
shell is in the correct directory for the Docker build process, to
avoid obscure bugs where the Docker context is not as expected.
This, again, currently relies on bash so may not extend to a fully
automated process.

alexandergeegds committed:
Ensure python logs as it goes

alexandergeegds committed:
Ignore all unexpected files

As with the replay-events folder, all unexpected files are now
ignored by Git. While this is slightly more effort for developers,
it prevents accidental inclusion of logs in git commits.

alexandergeegds committed:
Make it possible to override the output name

This is to allow an automated process to specify an exact filename
for use in a pipeline. This makes using the file in later steps of
the process simpler.

alexandergeegds committed:
Write out total number of downloaded requests

alexandergeegds committed:
Remove direct references to environment variables

These currently get printed to the console output during a pipeline
run, which given that they hold credentials is probably not a good
idea. Since we were relying on the shell builtin to verify the
variables had been defined, they now need explicit checks.

alexandergeegds committed:
Switch from bash to sh

Docker in docker does not have bash, so the helper scripts need to be
Bourne Shell friendly. The only thing that wasn't already compatible
was a reference to the current directory, which has been swapped out
for a hopefully-equivalent compatible version.

Note that the new version is taken from a SO answer. The code is
therefore available under the MIT licence already. However it does
require attribution, and guidance from [1] has been followed.

[1]: https://stackoverflow.blog/2009/06/25/attribution-required/

alexandergeegds committed:
Switch to Python 3-alpine image

This should reduce the download size considerably. (Seems to save
about 850MB with 3.7.) Given that dcind images for Concourse seem to
require downloading images each time, this is a also a substantial
time saving while running the helper script.

alexandergeegds committed:
Allow dry runs on replay

It is good to check something works. This allows runs to be 'dry-run'
by setting an environment variable to a non-empty value. This should
make testing initial setups easier.

alexandergeegds committed:
Allow use of a keyword for specifying start date

This allows automated runs to simply specify a relative date to
retrieve logs for.

alexandergeegds commented:
Going to temporarily switch the base to try and get GitHub to tidy up its diff... 🤞

chrisclayson commented on retrieve-logs/fetch_missing_matomo_requests.py:
```diff
@@ -64,47 +110,46 @@ def run_query(start_timestamp, end_timestamp):
+
+    period_start = datetime.utcfromtimestamp(start_datetime.replace(tzinfo=timezone.utc).timestamp())
+    total_written = 0
+    while period_start < end_datetime:
```
Should this be `<=` given you've done the `microseconds=-1` above?

alexandergeegds replied:
Yep, thanks for spotting. Have fixed in a new commit.

chrisclayson reviewed (changes requested):
See comment about the loop condition, you'll potentially miss 1 milliseconds worth of data ... if I've read it correctly!?

alexandergeegds committed:
Fix a loop conditional to not be strict

Since we are subtracting microseconds from our timestamps, the strict
inequality actually presents the possibility of omitting records from
the end timestamp. While the possibility is remote, we should include
everything if possible.

alexandergeegds committed:
Remove an unnecessary Docker container name

alexandergeegds committed:
Remove unused dependency

alexandergeegds committed:
Rename a shadowed variable

chrisclayson commented on retrieve-logs/retrieve.sh:
```diff
@@ -0,0 +1,37 @@
+
+set -u
+
+function be_in_directory {
```
Doesn't need changing, just wondering why we did this as function when it's only used once?

alexandergeegds replied:
1. I'm lazy. I have this as a snippet so I can copy/paste as necessary.
2. This feels like it fits the "self-documenting code" paradigm better. The alternative to me would be to do a 1-liner, but add a comment to explain what it's actually doing. (Because just reading shell scripts, I often have no clue... 😅 )

chrisclayson reviewed (dismissed):
👍

alexandergeegds committed:
Interpolate a variable correctly

It had been omitted when the variable was added. Now, if the variable
is updated, the error will too.

Co-authored-by: Chris Clayson <chris.clayson@digital.cabinet-office.gov.uk>

chrisclayson reviewed (dismissed):
👍

alexandergeegds committed:
Only download 5XX requests

Until the full consequences of replaying non-500s is understood, we
are omitting them in case there are dangers to attempting to replay
such traffic.

Co-authored-by: George Lund <george.lund@digital.cabinet-office.gov.uk>

alexandergeegds commented:
I have rebased the changes from #4 onto this branch, so there is now only the one PR to review. The only other recent change is restricting the Insights query to only download 5XX requests. Would appreciate a re-review / approval @galund @chrisclayson TIA
